pkg/collector/smi_collector.go

```diff
@@ -32,52 +39,70 @@ func (collector *SmiCollector) GetData() map[string]string {
+	if err != nil {
+		return err
+	}
+	for _, t := range allCrdsList.Items {
+		log.Printf("Test Names foo of %s", t.Name)
+	}
+	log.Printf("Test list output %s", allCrdsList)
+
 	// Filter to obtain a list of Smi CustomResourceDefinitions in the cluster
 	var smiCrdsList []string
-	for _, s := range allCrdsList {
-		if strings.Contains(s, "smi-spec.io") {
-			smiCrdsList = append(smiCrdsList, s)
+	for _, s := range allCrdsList.Items {
+		if strings.Contains(s.Name, "smi-spec.io") {
+			smiCrdsList = append(smiCrdsList, s.Name)
 		}
 	}
 	if len(smiCrdsList) == 0 {
 		return errors.New("cluster does not contain any SMI CRDs")
 	}
 
-	collectSmiCrds(collector, smiCrdsList)
+	collectSmiCrds(collector, smiCrdsList, apiextensionsClient)
 	return collectSmiCustomResourcesFromAllNamespaces(collector, smiCrdsList)
 }
 
-func collectSmiCrds(collector *SmiCollector, smiCrdsList []string) {
+func collectSmiCrds(collector *SmiCollector, smiCrdsList []string, apiextensionsClient *apiextensionsclientset.Clientset) {
 	for _, smiCrd := range smiCrdsList {
-		yamlDefinition, err := utils.RunCommandOnContainer("kubectl", "get", "crd", smiCrd, "-o", "yaml")
+		yamlDefinition, err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.Background(), smiCrd, metav1.GetOptions{})
 		if err != nil {
 			log.Printf("Skipping: unable to collect yaml definition of %s: %+v", smiCrd, err)
 		}
-		collector.data["smi/crd_"+strings.TrimSuffix(smiCrd, ".io")] = yamlDefinition
+		log.Printf("Test -----<> %v ", yamlDefinition)
+		collector.data["smi/crd_"+strings.TrimSuffix(smiCrd, ".io")] = yamlDefinition.String()
 	}
 }
 
 func collectSmiCustomResourcesFromAllNamespaces(collector *SmiCollector, smiCrdsList []string) error {
 	// Get all namespaces in the cluster
-	namespacesList, err := utils.GetResourceList([]string{"get", "namespaces", "-o", "jsonpath={..metadata.name}"}, " ")
+	// Creates the clientset
+	clientset, err := kubernetes.NewForConfig(collector.kubeconfig)
+	if err != nil {
+		return fmt.Errorf("getting access to K8S failed: %w", err)
+	}
+	namespacesList, err := clientset.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+
 	if err != nil {
 		return fmt.Errorf("collect SMI custom resources: unable to list namespaces in the cluster: %w", err)
 	}
 
-	for _, namespace := range namespacesList {
-		collectSmiCustomResourcesFromNamespace(collector, smiCrdsList, namespace)
+	for _, namespace := range namespacesList.Items {
+		collectSmiCustomResourcesFromNamespace(collector, smiCrdsList, namespace.Namespace)
 	}
 
 	return nil
 }
 
 func collectSmiCustomResourcesFromNamespace(collector *SmiCollector, smiCrdsList []string, namespace string) {
 	for _, smiCrdType := range smiCrdsList {
+		// TODO: Note - this part needs to be carefully tackle by folks who know exact working of this and expectations.
```

Comment:
@johnsonshi could you take a look at this?

Reply:
This can be refactored to use `smi-sdk-go`, but the problem is that these `go smi clients` cannot detect what the underlying version is. We *need* to run each `go smi client` for every `smi crd` and for every `smi crd version`.

For example, if we're trying to work with SMI Access CRDs, the code would be very verbose (to handle every version for every SMI CRD)


```go
import (
    ...
    ...
    smiAccessClient "github.com/servicemeshinterface/smi-sdk-go/pkg/gen/client/access/clientset/versioned"
    ...
)
```

```go


smiAccessList := smiAccessClient.AccessV1alpha1().TrafficTargets(dstPod.Namespace).List(context.TODO(), metav1.ListOptions{})
if len(smiAccessList) == 0 {
    // log.Printf("No SMI Access CRD V1 exists)
} else {
    // collect smi access crd info
}

smiAccessList = smiAccessClient.AccessV1alpha2().TrafficTargets(dstPod.Namespace).List(context.TODO(), metav1.ListOptions{})
if len(smiAccessList) == 0 {
    // log.Printf("No SMI Access CRD V2 exists)
} else {
    // collect smi access crd info
}

smiAccessList = smiAccessClient.AccessV1alpha3().TrafficTargets(dstPod.Namespace).List(context.TODO(), metav1.ListOptions{})
if len(smiAccessList) == 0 {
    // log.Printf("No SMI Access CRD V3 exists)
} else {
    // collect smi access crd info
}

```

What are your thoughts @SanyaKochhar?